components/feature/sendtab/src/main/java/mozilla/components/feature/sendtab/SendTabFeature.kt

```diff
@@ -0,0 +1,150 @@
+import mozilla.components.concept.push.Bus
+import mozilla.components.concept.push.PushService
+import mozilla.components.concept.sync.AccountObserver as SyncAccountObserver
+import mozilla.components.concept.sync.Device
+import mozilla.components.concept.sync.DeviceConstellation
+import mozilla.components.concept.sync.DeviceEvent
+import mozilla.components.concept.sync.DeviceEventsObserver
+import mozilla.components.concept.sync.DevicePushSubscription
+import mozilla.components.concept.sync.OAuthAccount
+import mozilla.components.concept.sync.TabData
+import mozilla.components.feature.push.AutoPushFeature
+import mozilla.components.feature.push.AutoPushSubscription
+import mozilla.components.feature.push.PushSubscriptionObserver
+import mozilla.components.feature.push.PushType
+import mozilla.components.service.fxa.manager.FxaAccountManager
+import mozilla.components.support.base.log.logger.Logger
+
+/**
+ * A feature that uses the [FxaAccountManager] to send and receive tabs with optional push support
+ * for receiving tabs from the [AutoPushFeature] and a [PushService].
+ *
+ * If the push components are not used, the feature can still function while tabs would only be
+ * received when the account is next synced.
+ *
+ * @param context Android application context for starting the (optional) push service.
+ * @param accountManager Firefox account manager.
+ * @param pushFeature The [AutoPushFeature] if that is setup for observing push events.
+ * @param pushService The optional [PushService] to manage start/stop of the service when
+ * an account is signed in/out.
+ * @param owner Android lifecycle owner for the observers. Defaults to the [ProcessLifecycleOwner]
+ * so that we can always observe events throughout the application lifecycle.
+ * @param autoPause whether or not the observer should automatically be
+ * paused/resumed with the bound lifecycle.
+ * @param onTabsReceived the callback invoked with new tab(s) are received.
+ */
+class SendTabFeature(
+    context: Context,
+    accountManager: FxaAccountManager,
+    pushFeature: AutoPushFeature? = null,
+    pushService: PushService? = null,
+    owner: LifecycleOwner = ProcessLifecycleOwner.get(),
+    autoPause: Boolean = false,
+    onTabsReceived: (Device?, List<TabData>) -> Unit
+) {
+    init {
+        val accountObserver = AccountObserver(context, pushFeature, pushService)
+        val pushObserver = PushObserver(accountManager)
+        val deviceObserver = DeviceObserver(onTabsReceived)
+
+        // Always observe the account for device events.
+        accountManager.registerForDeviceEvents(deviceObserver, owner, autoPause)
+
+        pushFeature?.apply {
+            registerForPushMessages(PushType.Services, pushObserver, owner, autoPause)
+            registerForSubscriptions(pushObserver, owner, autoPause)
+
+            // observe the account only if we have the push feature (service is optional)
+            accountManager.register(accountObserver, owner, autoPause)
```

Comment:
In the parameters to this class, both `pushFeature` and `pushService` are optional. However, here you're linking them together, since `accountObserver` is using `pushService`.. It's very easy to use the current API incorrectly, I think.

Another vote for dropping `pushService` from the parameters!

Reply:
The AccountObserver is only added if the feature is available, but the service is part of the nice-to-have part of this feature (see above comment).